### PR TITLE
User info

### PR DESCRIPTION
**Pull request by Ozzorany**
image

### CODE REVIEW

**Review by yarivg — APPROVED** [review]

File: src/app/services/firebase/firebase.service.ts [hunk 1]
```diff
@@ -33,6 +33,12 @@ export abstract class FirebaseService {
     });
   }
 
+  updateDocument<T extends { id: string }>(collection: CollectionType, data: T): Observable<T> {
+    return new Observable<T>(observer => {
```
Comment (ghost) [inline src/app/services/firebase/firebase.service.ts]: use 'from' operator like in delete

File: src/app/user/components/user-info/user-info.component.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      phoneNumberForm: new FormControl(null, [Validators.required, Validators.minLength(10)])
+    });
+
+    this.initializeUserDetails();
+    this.initializeRecaptcha();
+  }
+
+
+  initializeUserDetails(): void {
+    this.store.select(getUser).subscribe(user => {
```
Comment (ghost) [inline src/app/user/components/user-info/user-info.component.ts]: This is a memory leak. When you use subscribe you always need to unsubscribe.
Create an `onDestroy$' subject and use `takeUntil(onDestroy$)` 

check out this article:  - 3. Let’s Get Declarative With takeUntil
https://medium.com/angular-in-depth/the-best-way-to-unsubscribe-rxjs-observable-in-the-angular-applications-d8f9aa42f6a0

File: src/app/user/components/user-info/user-info.component.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+  get phoneNumberForm(): AbstractControl {
+    return this.editDetailsForm.get('phoneNumberForm');
+  }
+
+  dateChanged(date): void {
+    this.birthday = date.value;
+  }
+
+  private noWhitespaceValidator(control: FormControl): { whitespace: boolean } {
```
Comment (ghost) [inline src/app/user/components/user-info/user-info.component.ts:94]: This validator becomes repetitive.
Try to create a custom directive for it and use it in the sign-up form as well

File: src/app/services/firebase/firebase.service.ts (hunk 1 above)
Comment (ghost) [reply src/app/services/firebase/firebase.service.ts]: Why don't you use upsert? It should be the same

File: src/app/user/actions/user.actions.ts [hunk 4]
```diff
@@ -11,6 +11,10 @@ const userActionNames = {
   REGISTER: '[User] Register',
   REGISTER_SUCCESS: '[User] Register succeeded',
   REGISTER_FAIL: '[User] Register failed',
+  UPDATE: '[User] update',
```
Comment (ghost) [inline src/app/user/actions/user.actions.ts]: Add capital letter
update -> Update 
:)

File: src/app/user/components/user-info/user-info.component.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  saveDetails(): void {
+    this.user.firstName = this.firstNameForm.value;
+    this.user.lastName = this.lastNameForm.value;
+    this.user.email = this.emailForm.value;
+    this.user.birthday = firebase.firestore.Timestamp.fromDate(this.birthday);
+
+    if (this.user.phoneNumber === this.phoneNumberForm.value) {
+      this.store.dispatch(updateWithoutPhone({user: Object.assign({}, this.user)}));
```
Comment (ghost) [inline src/app/user/components/user-info/user-info.component.ts]: why not {user: this.user}
or
{user: {...this.user}}
?

File: src/app/user/components/user-info/user-info.component.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    } else {
+      this.user.phoneNumber = this.phoneNumberForm.value;
+      this.store.dispatch(update({user: Object.assign({}, this.user)}));
```
Comment (ghost) [inline src/app/user/components/user-info/user-info.component.ts]: same as above

File: src/app/user/services/users.service.ts [hunk 7]
```diff
@@ -1,27 +1,35 @@
...
+
   upsert$(data: User): Observable<User> {
     return this.firebaseService.upsertDocument<User>(this.collectionName, data);
   }
+
   delete$(data: User): Observable<void> {
     return this.firebaseService.deleteDocument<User>(this.collectionName, data);
   }
+
+  update$(data: User): Observable<User> {
```
Comment (ghost) [inline src/app/user/services/users.service.ts]: why not upsert?

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/app/services/firebase/firebase.service.ts (hunk 1 above)
Comment (Ozzorany) [reply src/app/services/firebase/firebase.service.ts]: removed

File: src/app/user/components/user-info/user-info.component.ts (hunk 2 above)
Comment (Ozzorany) [reply src/app/user/components/user-info/user-info.component.ts]: done.

File: src/app/user/components/user-info/user-info.component.ts (hunk 3 above)
Comment (Ozzorany) [reply src/app/user/components/user-info/user-info.component.ts:94]: I'll open another task and do that

File: src/app/user/actions/user.actions.ts (hunk 4 above)
Comment (Ozzorany) [reply src/app/user/actions/user.actions.ts]: done

File: src/app/user/components/user-info/user-info.component.ts (hunk 5 above)
Comment (Ozzorany) [reply src/app/user/components/user-info/user-info.component.ts]: done

File: src/app/user/components/user-info/user-info.component.ts (hunk 6 above)
Comment (Ozzorany) [reply src/app/user/components/user-info/user-info.component.ts]: done

File: src/app/user/services/users.service.ts (hunk 7 above)
Comment (Ozzorany) [reply src/app/user/services/users.service.ts]: removed

**Review by ghost — APPROVED** [review]

**Pull request merged**